### PR TITLE
Parameter golf jepa

### PR DESCRIPTION
**Pull request by ghost**
# JEPA-LM: First Joint Embedding Predictive Architecture for Text Compression

Gemini_Generated_Image_upkumfupkumfupku (1)

*"They said JEPA was only for vision. They were wrong."* — Yann LeCun, probably, upon seeing this PR

## Summary

We did the thing nobody asked for but everyone secretly wanted: we brought **JEPA (Joint Embedding Predictive Architecture)** to text compression. Yann can finally stop refreshing his timeline — someone actually used his architecture for language.

- **First application of JEPA to text compression / language model parameter golf**
- Reinterprets the U-Net transformer as a JEPA system with 256-dim latent projection, next-position predictor (MSE), and SIGReg anti-collapse regularizer
- Loss: `CE + 0.1*JEPA_MSE + 0.03*SIGReg` — JEPA components have **zero eval overhead** (they train the encoder to produce better representations, then politely step aside at inference)
- **val_bpb: 1.3355 mean** (seed 1337: 1.3294, seed 42: 1.3415, int8+zlib roundtrip)
- 19M params, 1212 lines, ~16.5MB artifact, 600s on 8xH100

## Architecture Innovation

While the rest of the world debates whether JEPA is the future of AI or just a very elaborate way to process images, we decided to settle the question by applying it to text compression. Spoiler: it works.

The encoder bottleneck is projected to a compact 256-dim latent space, and a predictor head learns to forecast the next position's embedding (with stop-gradient on targets, because we're not barbarians). **SIGReg** (Sketched Isotropic Gaussian Regularizer, adapted from vision/LeWorldModel) prevents representation collapse — because what's the point of a latent space if everything collapses to the same point?

```
Input tokens
    |
[Encoder: Layers 0-4] ---> skip connections stored
    |
  Bottleneck (h_enc)
    |--- jepa_proj ---> z (256-dim latent)
    |                    |--- predictor: z[t] -> z_pred[t+1]
    |                    |--- SIGReg: "please don't collapse"
    |
[Decoder: Layers 5-9] <--- skip connections consumed
    |
  Output logits (CE loss)
```

## Results

| Seed | Steps | val_bpb (int8+zlib) |
|------|-------|---------------------|
| 1337 | 9992  | 1.3294 |
| 42   | 10159 | 1.3415 |
| **Mean** | | **1.3355** |

Not beating SOTA, but we're not here to beat SOTA. We're here to prove that JEPA belongs in text compression. Consider this the architectural proof-of-concept that opens the door for future work.

## Adventures in torch.compile

Along the way we discovered that `torch.compile(fullgraph=True)` really does NOT appreciate you changing Python floats on a compiled model mid-training. The SIGReg weight decay caused a `FailOnRecompileLimitHit` that crashed our first run. Solution: `register_buffer` and keep the weight constant. Sometimes the best optimization is no optimization.

## Test plan

- [x] Smoke test (200 steps, 1 GPU)
- [x] Full training seed 1337 (8xH100, 600s)
- [x] Full training seed 42 (8xH100, 600s)
- [x] Int8+zlib quantization roundtrip validation
- [x] Line count < 1500 (1212 lines)
- [x] Made Yann LeCun proud (unverified)

---

*This submission was built in under 1 hour on 8xH100s. The JEPA components add only ~131K parameters (<1% of model) and have zero inference cost. We believe this is the first application of Joint Embedding Predictive Architecture to text compression.*

Best regards,

Daniel & Claude

### CODE REVIEW

**Comment by danielxmed** [comment]
> having CE in the loss defeats the purpose

No shit, Sherlock. Removing CE would also remove the ability to evaluate on bits-per-byte, submit to the leaderboard, or technically train a proper language model. The JEPA components are auxiliary, they shape the latent space during training and disappear at inference. CE is the actual objective that makes it a language model and lets it compete.
Would love to see your pure JEPA masterpiece that somehow magically scores on bpb without any CE. Genuinely curious how you'd pull that off.

Wait, you're the guy with the groundbreaking insight here, maybe you should show us how it's done!

**Comment by ShrvnKhmshn** [comment]
I agree that pure JEPA is likely ill-suited for this task, but given that using CE alone does better, I don't see what the point of this formulation is and how appropriate it is to be called JEPA (especially because of the decoder). Apologies if my prior comment seemed harsh.

**Comment by asuramaya** [comment]
im going to look into jepa

**Comment by MatoTeziTanka** [comment]
### Community Review — Parameter golf jepa

**Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

## Analysis ### Submission PR #1097 — "JEPA-LM: Joint Embedding Predictive Architecture for Text Compression" Head SHA: 5bed757868cf5b8508d09074ad42dd12e96abdaa Author: danielxmed Reported val_bpb: 1.3355 ### Check 1: ILLEGAL n-gram family bug (target XOR'd into hash key / BigramHash) CLEAR. No n-gram, bigram, hash, or XOR constructs anywhere in the submission. No lookup tables indexed by target tokens at train time. Not applicable. ### Check 2: ILLEGAL Pre-Quant TTT (multi-epoch on val_tokens without score-first) CLEAR. `val_tokens` is used exclusively in `eval_val()` (lines 233–285) under `torch.inference_mode()` with `model.eval()`. This function is called twice: - Line 1065: periodic validation inside the training loop (read-only, no gradient). - Line 1188: post-quantization round-trip check (also read-only, no gradient). No backward pass, optimizer step, or fine-tuning loop touches `val_tokens` at any point. ### Check 3: LEGAL score-first TTT (PR #1413 pattern, is_last_chunk guard) NOT PRESENT. There is no TTT of any kind — no score-first guard, no `is_last_chunk`, no adaptation on val data. Not applicable; absence is fine for PURE_NEURAL_CLEAN. ### Check 4: HOLD scored-region SLOT NOT PRESENT. No scored-region slot exploitation detected. ### Check 5: Clean pure neural CONFIRMED. The submission is a straightforward transformer variant (U-Net / JEPA encoder-decoder) trained only on `train_loader` batches. The JEPA auxiliary loss (MSE on latent embeddings, line 792) and SIGReg regularizer (line 795) are computed exclusively during training mode (`if not self.training: return ce_loss` at line 787). No val-data leakage, no hash tricks, no TTT, no external lookup corruption. Architecture novelty (JEPA projection + predictor, ~131K extra params) is...

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the audit — this looks like a clean pure-neural submission.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*
